### PR TITLE
Fix inconsistency on failure updating

### PR DESCRIPTION
**Pull request by josenavas**
~~This PR depends on #1050, review and merge that one first.~~

Main motivation of the PR: the `update_category` function can leave the DB in an inconsistent state because the updates to the samples are done independently, so if one fails, the previous ones are already modified. By moving the `__setitem__` queries to a function that adds them to a queue, we can use that function in `update_category` to perform everything in a single transaction block.

Also, in a previous PR, @squirrelo commented that the way types in the error were checked was probably not correct (for some reason I received that notification from github today). I tested it and @squirrelo was right, so I fixed it here. It is less elegant, but at least is correct.

### CODE REVIEW

**Comment by squirrelo** [comment]
One small comment.

File: qiita_db/metadata_template/base_metadata_template.py [hunk 1]
```diff
@@ -363,27 +364,51 @@ def __setitem__(self, column, value):
...
             # column type or something else
-            column_type = conn_handler.execute_fetchone(
-                """SELECT data_type
-                   FROM information_schema.columns
-                   WHERE column_name=%s AND table_schema='qiita'
-                """, (column,))[0]
-            value_type = type(value).__name__
-
-            if column_type != value_type:
+            if "invalid input syntax for" in str(e):
```
Comment (squirrelo) [inline qiita_db/metadata_template/base_metadata_template.py]: Something like [this](https://github.com/biocore/qiita/blob/master/qiita_db/search.py#L277) from search.py would probably be cleaner, since it directly connects the python datatype to the SQL datatype and the invalid syntax could be for other reasons besides type.

File: qiita_db/metadata_template/base_metadata_template.py [hunk 2]
```diff
@@ -1101,12 +1126,44 @@ def update_category(self, category, samples_and_values):
...
             sample = self[k]
-            sample[category] = v
+            sample.add_setitem_queries(category, v, conn_handler, queue_name)
+
+        try:
+            conn_handler.execute_queue(queue_name)
+        except QiitaDBExecutionError as e:
+            # catching error so we can check if the error is due to different
+            # column type or something else
+            if "invalid input syntax for" in str(e):
```
Comment (squirrelo) [inline qiita_db/metadata_template/base_metadata_template.py]: same as above

**Comment by squirrelo** [comment]
One comment.

**Comment by josenavas** [comment]
This should be ready for final review/merge!

**Comment by josenavas** [comment]
ping @biocore/qiita-dev

File: qiita_db/metadata_template/base_metadata_template.py [hunk 3]
```diff
@@ -313,8 +315,8 @@ def handler(x):
                            " in template %d" %
                            (key, self._id, self._md_template.id))
 
-    def __setitem__(self, column, value):
-        r"""Sets the metadata value for the category `column`
+    def add_setitem_queries(self, column, value, conn_handler, queue):
+        """Adds the SQL queries needed to set and item to the queue
```
Comment (ElDeveloper) [inline qiita_db/metadata_template/base_metadata_template.py]: Can you rephrase this docstring, I don't think it makes sense.

**Comment by ElDeveloper** [comment]
Looks good, but I think this might be missing some tests?

File: qiita_db/metadata_template/base_metadata_template.py (hunk 3 above)
Comment (josenavas) [inline qiita_db/metadata_template/base_metadata_template.py]: Done

**Comment by josenavas** [comment]
Thanks @ElDeveloper 
I've added tests!

**Comment by ElDeveloper** [comment]
Thanks @josenavas!
